src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs

```diff
@@ -225,46 +224,91 @@ private static (SyntaxTriviaList sameLine, SyntaxTriviaList endOfLine) SplitTrai
         {
-            foreach (var nodeOrToken in declarationSyntax.Parent.ChildNodesAndTokens().Reverse())
+            // We are going to step through the statements starting with the trigger variable's declaration.
+            // We will track when new locals are declared and when they are used. To determine the last
+            // statement that we should surround, we will walk through the locals in the order they are declared.
+            // If the local's declaration index falls within the last variable usage index, we will extend
+            // the last variable usage index to include the local's last usage.
+
+            // Take all the statements starting with the trigger variable's declaration.
+            var statementsAfterDeclaration = declarationSyntax.Parent.ChildNodesAndTokens()
+                .Select(nodeOrToken => nodeOrToken.AsNode())
+                .OfType<TStatementSyntax>()
+                .SkipWhile(node => node != declarationSyntax)
+                .ToArray();
+
+            // List of local variables that will be in the order they are declared.
+            var localVariables = new List<ISymbol>();
+
+            // Map a symbol to an index into the statementsAfterDeclaration 
+            var variableDeclarationIndex = PooledDictionary<ISymbol, int>.GetInstance();
+            var lastVariableUsageIndex = PooledDictionary<ISymbol, int>.GetInstance();
+
+            for (var statementIndex = 0; statementIndex < statementsAfterDeclaration.Length; statementIndex++)
             {
-                var node = (TStatementSyntax)nodeOrToken.AsNode();
-                if (node is null)
+                var node = statementsAfterDeclaration[statementIndex];
+
+                // Update the last usage index for any variable referenced in the statement.
+                var referencedVariables = GetReferencedLocalVariables(node, localVariables, semanticModel, syntaxFactsService, cancellationToken);
+                foreach (var referencedVariable in referencedVariables)
                 {
-                    continue;
+                    lastVariableUsageIndex[referencedVariable] = statementIndex;
                 }
 
-                if (node == declarationSyntax)
+                // Add new variables that were declared in the statement
+                var declaredVariables = semanticModel.GetAllDeclaredSymbols(node, cancellationToken);
+                foreach (var declaredVariable in declaredVariables)
                 {
-                    break; // Ignore the declaration and usages prior to the declaration
+                    variableDeclarationIndex[declaredVariable] = statementIndex;
+                    lastVariableUsageIndex[declaredVariable] = statementIndex;
+                    localVariables.Add(declaredVariable);
                 }
+            }
 
-                if (ContainsReference(node, localVariable, semanticModel, syntaxFactsService, cancellationToken))
+            var lastUsageIndex = 0;
```

Comment:
Is this right? I imagine this breaks if you have a statement in between the variable for the usings and the next variable assignment. 

```C# 
var a = MakeSomeDisposable();
a.DoSomething();
var b = a.GetB();
```

Does the above get correctly handled?

Reply:
Talked offline, this is correct

Reply:
Added more comments to improve clarity.